### PR TITLE
add get_main_module api

### PR DESCRIPTION
**Pull request by scxiao**

### CODE REVIEW

File: src/include/migraphx/program.hpp [hunk 1]
```diff
@@ -139,6 +141,8 @@ struct program
     friend bool operator==(const program& x, const program& y);
     friend bool operator!=(const program& x, const program& y) { return !(x == y); }
 
+    module& get_main_module() { return *this; }
```
Comment (pfultz2) [inline src/include/migraphx/program.hpp:144]: Can this be added to our python/C API?

Comment (scxiao) [reply src/include/migraphx/program.hpp:144]: Thanks. Looks like I forgot. just added. will add corresponding unit tests tomorrow.

Comment (pfultz2) [reply src/include/migraphx/program.hpp:144]: This should return `module*` not `module&`.

**Review by kahmed10 — APPROVED** [review]

File: src/api/api.cpp [hunk 2]
```diff
@@ -616,11 +628,57 @@ migraphx_shapes_get(const_migraphx_shape_t* out, migraphx_shapes_t shapes, size_
...
+    return migraphx::try_([&] { destroy((module)); });
+}
+
+extern "C" migraphx_status migraphx_module_create(migraphx_module_t* module)
+{
+    return migraphx::try_(
+        [&] { *module = object_cast<migraphx_module_t>(allocate<migraphx::module>()); });
+}
+
+extern "C" migraphx_status migraphx_module_compile(migraphx_module_t module,
```
Comment (pfultz2) [inline src/api/api.cpp:642]: A module shouldn't have a compile function.

File: src/api/migraphx.py [hunk 3]
```diff
@@ -163,8 +163,19 @@ def shapes(h):
              returns='const migraphx::shape&')
 
 
+@auto_handle
+def module(h):
+    h.constructor('create')
```
Comment (pfultz2) [inline src/api/migraphx.py:168]: I dont think we should allow constructing a module. A module should only come from the program(ie `get_main_module`), and the program will own the module, so there is no need to expose a `destroy` or `create` function either.

File: src/include/migraphx/program.hpp (hunk 1 above)
Comment (scxiao) [reply src/include/migraphx/program.hpp]: fixed

File: src/api/migraphx.py (hunk 3 above)
Comment (pfultz2) [reply src/api/migraphx.py:168]: So PR#673 adds the ability to not generate the `destroy` method. It just needs to be changed to `@auto_handle(ref=True)`.

Comment (scxiao) [reply src/api/migraphx.py:168]: For the c++ api, looks like handle_base struct needs a destroy function, how can we handle that? definition of the handle base is 
template <class T, class D, D Deleter>
struct handle_base
{}

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**